# Introduction

This tutorial will teach you how to access real life data and understand it visually.

For this article, we employ one of the data sets available in muenchen.de, containing information about Oktoberfest since 1985 until 2019. Although most of the years are not available (Oktoberfest started in 1810!), the data set provides valuable insights about how the festival evolved over the last 30 years.

The original data has been taken from here: https://www.opengov-muenchen.de/dataset/oktoberfest (no need to go there now)


So, let's jump into it!

# Libraries to Use

To evaluate the dataset, we use three different Python libraries: Pandas, Matplotlib, and Plotly.

- [Pandas](https://pandas.pydata.org/index.html) is a Python open source library for data science that allows us to easily work with structured data, such as **csv files, SQL tables, or Excel spreadsheets**. It provides tools for reading and writing data in different formats, carrying out exploratory analysis, and cleaning data (reshaping data sets, handling missing data, or merging data sets).

- [Matplotlib](https://matplotlib.org/) is a Python 2D plotting library that can be used in Python scripts, Jupyter notebooks, and IPython shells, among other environments, producing high quality figures. Matplotlib offers a wide range of visualizations such as histograms, bar chart, scatter plots, box plots, or pie charts.

- [Plotly Python](https://plot.ly/python/) graphing library provides as Matplotlib a wide range of visualizations. But, unlike Matplotlib the visualizations provided by Plotly are interactive and more intuitive. 😉
After this overview about the libraries used in this analysis, we are ready to start the next step: exploratory data analysis and data cleaning.

After this overview of the libraries used in this analysis, we are ready to load the dataset.

### Working with directories
This module provides a portable way of using operating system dependent functionality, such as retrieving information about paths and files



In [ ]:
import os

### to see where we are currently working in 
print(os.getcwd())

# os.path.join
# os.listdir()
# etc.


/content


### Working with paths
For a more intuitive approach of working with local files we use the module [pathlib](https://docs.python.org/3/library/pathlib.html) which we will discuss in a later exercise in more detail.



In [ ]:
# importing the pathlib module - a handy library for working with the local file system in an object oriented way
from pathlib import Path

# creating a path object of our data directory within the mounted Google Drive
oktoberfest_path = Path('/content/')

print(oktoberfest_path)


/content


In [ ]:
### to download the .csv file
!pip install wget

import wget

wget.download('https://www.opengov-muenchen.de/dataset/8d6c8251-7956-4f92-8c96-f79106aab828/resource/e0f664cf-6dd9-4743-bd2b-81a8b18bd1d2/download/oktoberfestgesamt19852019.csv')


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9675 sha256=164c6e589f5b77687e7d4cf4ad7d4c81e9757932e4828a36174c4ae01c5f45e9
  Stored in directory: /root/.cache/pip/wheels/a1/b6/7c/0e63e34eb06634181c63adacca38b79ff8f35c37e3c13e3c02
Successfully built wget


'oktoberfestgesamt19852019.csv'

In [ ]:
# print all the files and directories in our data path
print('\nFiles in our oktoberfest path:')
for file in oktoberfest_path.iterdir():
  print(file)


Files in our oktoberfest path:
/content/.config
/content/oktoberfestgesamt19852019.csv
/content/sample_data


Great... .csv files found. 

CSV stands for comma-separated values. This is a text-based file format that stores tabular data with each line being a row and columns being separated by commas. Just like a minimalistic Excel file without formating and only values. ([More info on wikipedia](https://en.wikipedia.org/wiki/Comma-separated_values))

Let's take the 2019 version and save the path to a variable:



In [ ]:
# create another pathlib object with the path to the csv file
oktoberfest_csv_path = oktoberfest_path / 'oktoberfestgesamt19852019.csv'

# make sure this is a file
assert oktoberfest_csv_path.is_file(), 'Error oktoberfest csv file not found!'
print('Oktoberfest csv file found:', oktoberfest_csv_path.is_file())

Oktoberfest csv file found: True


For convenience we load the CSV data into a Pandas data frame using the [pandas.read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function and visualize the first 5 rows using the [pandas.DataFrame.head](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html) method.

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
# make plots look prettier
plt.style.use('ggplot')

# Load data into a pandas dataframe
df_oktoberfest = pd.read_csv(oktoberfest_csv_path)

# Print first 5 rows of the dataframe
df_oktoberfest.head(5)

jahr  dauer  besucher_gesamt  besucher_tag  bier_preis  bier_konsum  \
0  1985     16              7.1           444        3.20        54541   
1  1986     16              6.7           419        3.30        53807   
2  1987     16              6.5           406        3.37        51842   
3  1988     16              5.7           356        3.45        50951   
4  1989     16              6.2           388        3.60        51241   

   hendl_preis  hendl_konsum  
0         4.77        629520  
1         3.92        698137  
2         3.98        732859  
3         4.19        720139  
4         4.22        775674

The data set contains 8 columns: (1) year, (2) duration, (3) total visitors in million, (4) visitors per day in thousands, (5) price of beer, (6) consumption of beer in hektoliters, (7) price of chicken, (8) consumption of chicken.

We can also get the last 5 rows using the [pandas.DataFrame.tail](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.tail.html) method.

In [ ]:
# print the last 5 rows of the dataframe
df_oktoberfest.tail(5)

jahr  dauer  besucher_gesamt  besucher_tag  bier_preis  bier_konsum  \
30  2015     16              5.9           369       10.27        75134   
31  2016     17              5.6           329       10.57        66231   
32  2017     18              6.2           344       10.87        77836   
33  2018     16              6.3           394       11.30        78705   
34  2019     16              6.3           394       11.71        78502   

    hendl_preis  hendl_konsum  
30        10.75        503510  
31        11.00        366876  
32        11.41        466747  
33        11.67        436492  
34        12.15        434998

Let's get a feeling for the data and print some statistics with the [pandas.DataFrame.describe](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) method.

Of course you could also calculate single statistics like .mean(), .std() etc..

In [ ]:
# generate and print descriptive statistics of the dataframe
df_oktoberfest.describe()

jahr      dauer  besucher_gesamt  besucher_tag  bier_preis  \
count    35.000000  35.000000        35.000000     35.000000   35.000000   
mean   2002.000000  16.285714         6.314286    388.171429    6.870571   
std      10.246951   0.621735         0.390432     26.337723    2.581078   
min    1985.000000  16.000000         5.500000    329.000000    3.200000   
25%    1993.500000  16.000000         6.050000    369.000000    4.800000   
50%    2002.000000  16.000000         6.400000    394.000000    6.750000   
75%    2010.500000  16.000000         6.500000    406.000000    8.940000   
max    2019.000000  18.000000         7.100000    444.000000   11.710000   

        bier_konsum  hendl_preis   hendl_konsum  
count     35.000000    35.000000      35.000000  
mean   62223.371429     7.592000  571920.714286  
std    10229.542414     2.621272  123620.418372  
min    48698.000000     3.920000  351705.000000  
25%    53214.500000     5.295000  482005.000000  
50%    61163.000000     8.120000  521872.000000  
75%    70203.000000     9.715000  689689.500000  
max    79225.000000    12.150000  807710.000000

In python, a dataframe is very similar to a dictionary. To get single columns you can use the column names.

In [ ]:
# let's print the column names first
print(df_oktoberfest.columns)

Index(['jahr', 'dauer', 'besucher_gesamt', 'besucher_tag', 'bier_preis',
       'bier_konsum', 'hendl_preis', 'hendl_konsum'],
      dtype='object')


In [ ]:
# get the column for the beer price via the dictionary syntax:
df_oktoberfest['bier_preis']

0      3.20
1      3.30
2      3.37
3      3.45
4      3.60
5      3.77
6      4.21
7      4.42
8      4.71
9      4.89
10     5.15
11     5.24
12     5.45
13     5.60
14     5.80
15     6.35
16     6.47
17     6.75
18     6.75
19     7.02
20     7.15
21     7.42
22     7.85
23     8.25
24     8.53
25     8.79
26     9.09
27     9.41
28     9.78
29     9.98
30    10.27
31    10.57
32    10.87
33    11.30
34    11.71
Name: bier_preis, dtype: float64

In [ ]:
# You can also use a namespace to access the same column instead of the dictionary syntax:
df_oktoberfest.bier_preis

0      3.20
1      3.30
2      3.37
3      3.45
4      3.60
5      3.77
6      4.21
7      4.42
8      4.71
9      4.89
10     5.15
11     5.24
12     5.45
13     5.60
14     5.80
15     6.35
16     6.47
17     6.75
18     6.75
19     7.02
20     7.15
21     7.42
22     7.85
23     8.25
24     8.53
25     8.79
26     9.09
27     9.41
28     9.78
29     9.98
30    10.27
31    10.57
32    10.87
33    11.30
34    11.71
Name: bier_preis, dtype: float64

In [ ]:
# Let's calculate the lowest price for one liter of beer recorded .. :-/
df_oktoberfest.bier_preis.min()

3.2

So how do we access a specific row?

Let's try the [pandas.DataFrame.iloc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html) function to view specific years. This function takes the index of each row as the argument, and outputs the corresponding row.

In [ ]:
df_oktoberfest.iloc[[0]]

You can get multiple rows like this.


In [ ]:
df_oktoberfest.iloc[[0,1,2]]

Now, let's try to do some data anlaysis. One of the biggest change in German economy was the conversion from Deutsche Mark to Euro. This was roughly around 1999 to 2000. Let's check the data from that years.

In [ ]:
df_oktoberfest.iloc[[12,13,14,15,16,17],:]

From 1999 to 2000 there is a sudden increase in the beer and chicken price. This can be attributed to the conversion of Deutsche Mark to Euro as the official currency of Germany.

We set year as the index of the data frame. To easily access the information of different years.

In [ ]:
# Set year as the index of the data frame
df_oktoberfest.set_index('jahr', inplace=True)


# Print first 5 rows of the dataframe
df_oktoberfest.head()

# Analysing the Data

Exploratory data analysis allow us to get a feeling about the data set and to get the data set ready to easily draw conclusions using it.
Let's try to answer some common questions.

### How did beer price change over the last 34 years? Is there a tendency?

Pandas provides a very convenient method for plotting our data frame: [pandas.DataFrame.plot](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.plot.html)

In [ ]:
df_oktoberfest.plot()

This includes all the columns and is not really useful.

Let's specify more argument regarding the columns to include (y) and some formatting ([for more details check the pandas help](https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html)):

In [ ]:
# Plot beer price
df_oktoberfest.plot(kind='line', y=['bier_preis'], figsize=(15,8), linewidth=3, fontsize=16)

# Title, labels and legend
plt.legend(labels=['Beer price'],fontsize=16)
plt.xlabel('Year',fontsize=16)
plt.ylabel('Price (Euro)',fontsize=16)
plt.title('History of prices from 1985 until 2018',fontsize=20)

As we can see from the plot, the price of beer have increased steadily from 1985 until 2019,
Matter of fact, the price in 2019 is around three times more than in 1985.

Try to answer the following question yourself.

### How did chicken price change over the last 34 years? 

In [ ]:
# Plot chicken price

### ENTER YOUR CODE ###


# Title, labels and legend

### ENTER YOUR CODE ###

In [ ]:
# Price of a beer and a chicken in 1985
print(df_oktoberfest.loc[1985,['bier_preis','hendl_preis']].sum())


# Price of a beer and a chicken in 2018
print(df_oktoberfest.loc[2018,['bier_preis','hendl_preis']].sum())


So, you can see that a person in 1985 could enjoy a meal at the Oktoberfest for only 7.97 euros. Three decades later, we need almost three times more money (22.97 euros) to enjoy a meal at the festival.

Here we visualize a interactive plot of the prices

In [ ]:
import plotly.graph_objects as go

# Interactive plots of the price of beer and chicken
year = df_oktoberfest.index
beer_price = df_oktoberfest.bier_preis
chicken_price = df_oktoberfest.hendl_preis

fig = go.Figure()

fig.add_trace(go.Scatter(x=year, y=chicken_price,
                    mode='lines+markers',
                    name='Chicken'))

fig.add_trace(go.Scatter(x=year, y=beer_price,
                    mode='lines+markers',
                    name='Beer'))

# Title, labels, and background
fig.update_layout(title='History of prices from 1985 until 2019',
                   xaxis_title='Year',
                   yaxis_title='Price (Euro)',
                   plot_bgcolor='white')

### Beer Consumption

Let's take a look at the beer consumption over the years.



In [ ]:
# Plot beer consumption 
df_oktoberfest.plot(y='bier_konsum', figsize=(15,8), linewidth=3, fontsize=16, color='maroon', label='_nolegend_')

# Years Landwirtschaftsfest
years_landwirtschaftsfest = [1987, 1990, 1993, 1996, 2000, 2004, 2008, 2012, 2016]

# Plot vertical line, indicating the celebration of the Landwirtschaftsfest. 
for year in years_landwirtschaftsfest:
    plt.axvline(x=year, color='blue', alpha=0.3)


# Annotations - celebration of the landwirtschaftsfest and terrorist attacks  
plt.annotate('Vertical lines indicate the celebration of the Landwirtschaftsfest', xy=(2008, 75000), xytext=(1985, 79000), fontsize=15,
            arrowprops=dict(facecolor='black', shrink=0.05),bbox = dict(boxstyle="round", fc="w",ec="maroon"),
            )
plt.annotate('Terrorist attacks 9/11', xy=(2001, df_oktoberfest.loc[2001,'bier_konsum']), xytext=(2004, 52500), fontsize=15,
            arrowprops=dict(facecolor='black', shrink=0.05),bbox = dict(boxstyle="round", fc="w",ec="maroon"),
            )
plt.annotate('Terrorist attacks Paris and Nice', xy=(2016, df_oktoberfest.loc[2016,'bier_konsum']), xytext=(2005, 59000), fontsize=15,
            arrowprops=dict(facecolor='black', shrink=0.05),bbox = dict(boxstyle="round", fc="w",ec="maroon"),
            )

# Title and labels
plt.xlabel('Year',fontsize=16)
plt.ylabel('Bier consumption (Hectoliters)',fontsize=16)
plt.title('Consumption of bier',fontsize=20)

As we can observe, the consumption of beer presents a growing trend. The vertical lines indicate the years than the Oktoberfest coincided with the celebration of the Landswirtschaftsfest. This festival is celebrated every four years since 1996 (before every 3) at Theresienwiesen at the same time that Oktoberfest. We can see that small decreases in 2004, 2008, and 2012 could be related to the celebration of this festival.

On the contrary, the significant decreases of consumption of beer in 2001 and 2016 could be related to the September 11 terrorist attacks in 2001, the November 2015 Paris terrorist attacks, and the 2016 Nice truck attack. The number of visitors decreased in those years (2001 and 2016), showing the lowest values since 1985.


In [ ]:
# Lowest number of visitors since 1985 in 2001,2016, and 1988. 
df_oktoberfest.besucher_gesamt.sort_values().head(3)

### Chicken Consumption

Now, you try the same for the chicken consumption and see for yourself if we have the same trend as the beer consumption.

You would need to do only minor changes to the code of beer consumption.

In [ ]:

# Plot chicken consumption 

### ENTER YOUR CODE ###


# Years Landwirtschaftsfest
years_landwirtschaftsfest = [1987, 1990, 1993, 1996, 2000, 2004, 2008, 2012, 2016]

# Plot vertical line, indicating the celebration of the Landwirtschaftsfest. 
for year in years_landwirtschaftsfest:
    plt.axvline(x=year, color='blue', alpha=0.3)


# Annotations - celebration of the landwirtschaftsfest and terrorist attacks  

### ENTER YOUR CODE ###


# Title and labels
plt.xlabel('Year',fontsize=16)
plt.ylabel('Hendl consumption',fontsize=16)
plt.title('Consumption of chicken',fontsize=20)

### Number of Visitors

Finally let's take a look at the total number of visitors from 1985 to 2019.

In [ ]:
# Plot total number of visitors
df_oktoberfest.plot(y='besucher_gesamt', figsize=(13,8), marker='o', linewidth=2, fontsize=16, color='darkblue', label='_nolegend_')


# Title and labels
plt.xlabel('Year',fontsize=16)
plt.ylabel('Number of visitors (Milions)',fontsize=16)
plt.title('Total number of visitors',fontsize=20)

We can easily analyze the number of visitors by using the above plot. As we can observe, the number of visitors does not follow any trend, ranging from 5.5 to 7.1. So, the consumption of beer is not increasing because the festival has more visitors, but because the people drink more. 


# References
- https://towardsdatascience.com/oktoberfest-quick-analysis-using-pandas-matplotlib-and-plotly-79778367a4c
- https://www.opengov-muenchen.de/dataset/oktoberfest
- https://pandas.pydata.org/index.html
- https://matplotlib.org/
- https://plot.ly/python/